libslide/src/evaluator_rules.rs

```diff
@@ -109,36 +112,64 @@ impl Transformer<Expr, Expr> for BuiltRule {
     ///   "$x + 0 -> $x".try_apply("x + 1")  // None
     ///   "$x + 0 -> $x".try_apply("x")      // None
     ///
-    fn transform(&self, target: Expr) -> Expr {
-        // First, apply the rule recursively on the target's subexpressions.
-        let partially_transformed = match target {
-            expr @ Expr::Const(_) => expr,
-            var @ Expr::Var(_) => var,
-            Expr::BinaryExpr(binary_expr) => BinaryExpr {
-                op: binary_expr.op,
-                lhs: self.transform(*binary_expr.lhs).into(),
-                rhs: self.transform(*binary_expr.rhs).into(),
-            }
-            .into(),
-            Expr::UnaryExpr(unary_expr) => UnaryExpr {
-                op: unary_expr.op,
-                rhs: self.transform(*unary_expr.rhs).into(),
+    fn transform(&self, target: Rc<Expr>) -> Rc<Expr> {
+        fn fill(cache: &mut HashMap<u64, Rc<Expr>>, t: Rc<Expr>, r: Rc<Expr>) -> Rc<Expr> {
+            Rc::clone(cache.entry(hash(t.as_ref())).or_insert_with(|| r))
+        }
```

Comment:
why do we need to use Rc::clone. Can't we just originally create an Rc pointer instead of allocating a new one?

Reply:
This is the great thing about `Rc`! Cloning gives a new reference to the inner pointer, not actually cloning the inner data.